### PR TITLE
[ADR] ADR-052: Tendermint Mode

### PR DESCRIPTION
**Pull request by dongsam**
### ADR 052: Tendermint Mode

- Fullnode mode : fullnode mode does not have any capability to participate on consensus
- Validator mode : this mode is exactly same as existing state machine behavior. sync without voting on consensus, and participate consensus when fully synced
- Seed mode : lightweight seed mode only for maintain an address book, p2p like [TenderSeed](https://gitlab.com/polychainlabs/tenderseed)

Separate ADR Tendermint Mode from ADR-051 https://github.com/tendermint/tendermint/pull/4262

* [x] Referenced an issue explaining the need for the change
* [x] Updated all relevant documentation in docs
* [ ] Updated all code comments where relevant
* [ ] Wrote tests
* [ ] Updated CHANGELOG_PENDING.md

### CODE REVIEW

File: docs/architecture/adr-052-tendermint-mode.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+## Changelog
+
+* 27-11-2019: Initial draft from ADR-051
+* 13-01-2020: Separate ADR Tendermint Mode from ADR-051
+
+## Context
+
+- Fullnode mode : fullnode mode does not have any capability to participate on consensus
+- Validator mode : this mode is exactly same as existing state machine behavior. sync without voting on consensus, and participate consensus when fully synced
+- Seed mode : lightweight seed mode only for maintain an address book, p2p like [TenderSeed](https://gitlab.com/polychainlabs/tenderseed)
```
Comment (melekes) [inline docs/architecture/adr-052-tendermint-mode.md]:
```suggestion
- Seed mode : lightweight seed mode maintaining an address book, p2p like [TenderSeed](https://gitlab.com/polychainlabs/tenderseed)
```

File: docs/architecture/adr-052-tendermint-mode.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
+# ADR 052: Tendermint Mode
+
+## Changelog
+
+* 27-11-2019: Initial draft from ADR-051
+* 13-01-2020: Separate ADR Tendermint Mode from ADR-051
+
+## Context
+
+- Fullnode mode : fullnode mode does not have any capability to participate on consensus
```
Comment (marbar3778) [inline docs/architecture/adr-052-tendermint-mode.md]:
```suggestion
- Fullnode mode: fullnode mode does not have the capability to become a validator. 
```

I believe participating in consensus is also passing consensus messages around correct?

File: docs/architecture/adr-052-tendermint-mode.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+- Seed mode : lightweight seed mode maintaining an address book, p2p like [TenderSeed](https://gitlab.com/polychainlabs/tenderseed)
+
+## Decision
+
+We would like to suggest a simple Tendermint mode abstraction
+
+- Which reactor, component to include for each node
+    - fullnode *(default)*
+        - switch, transport
+        - all reactors
```
Comment (marbar3778) [inline docs/architecture/adr-052-tendermint-mode.md]: mempool are considered reactors to my knowledge, maybe better as
```suggestion
        - reactors
          - mempool
          - consensus
          - ...
```
then listing all the reactors involved

File: docs/architecture/adr-052-tendermint-mode.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        - mempool, state
+        - rpc
```
Comment (marbar3778) [inline docs/architecture/adr-052-tendermint-mode.md]:
```suggestion
        - rpc (safe connections only)
```
?

File: docs/architecture/adr-052-tendermint-mode.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        - only pexReactor
+- Configuration, cli command
+    - We would like to suggest by introducing `mode` parameter in `config.toml` and cli
+    - `mode = "{{ .BaseConfig.Mode }}"` in `config.toml`
+    - `tendermint node --mode validator`  in cli
+    - fullnode | validator | seed (default: "fullnode")
+- RPC modification
+    - `host:26657/status`
+        - return empty `validator_info` when fullnode mode
+        - add return `mode` of this node
```
Comment (marbar3778) [inline docs/architecture/adr-052-tendermint-mode.md]: this issue may have some thoughts on introducing which mode the node is in. https://github.com/tendermint/tendermint/issues/2817

**Comment by marbar3778** [comment]
this will all be under one binary correct?

File: docs/architecture/adr-052-tendermint-mode.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+## Context
+
+- Fullnode mode : fullnode mode does not have any capability to participate on consensus
+- Validator mode : this mode is exactly same as existing state machine behavior. sync without voting on consensus, and participate consensus when fully synced
+- Seed mode : lightweight seed mode maintaining an address book, p2p like [TenderSeed](https://gitlab.com/polychainlabs/tenderseed)
+
+## Decision
+
+We would like to suggest a simple Tendermint mode abstraction
```
Comment (marbar3778) [inline docs/architecture/adr-052-tendermint-mode.md]:
```suggestion
We would like to suggest a simple Tendermint mode abstraction. These modes will live under one binary, and when initializing a node the user will be able to specify which node they would like to create.
```

maybe something like this? just to expand a bit more

**Comment by dongsam** [comment]
> this will all be under one binary correct?

Yes, it will all be under one binary 
Thank you for adding details.

File: docs/architecture/adr-052-tendermint-mode.md (hunk 5 above)
Comment (dongsam) [reply docs/architecture/adr-052-tendermint-mode.md]: how about adding only line `return empty validator_info when fullnode mode` without line `add return mode of this node`?

File: docs/architecture/adr-052-tendermint-mode.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+We would like to suggest a simple Tendermint mode abstraction. These modes will live under one binary, and when initializing a node the user will be able to specify which node they would like to create.
+
+- Which reactor, component to include for each node
+    - fullnode *(default)*
+        - switch, transport
+        - reactors
+          - mempool
+          - consensus
+          - ...
```
Comment (marbar3778) [inline docs/architecture/adr-052-tendermint-mode.md]:
```suggestion
          - evidence
          - blockchain
          - p2p/pex
```

File: docs/architecture/adr-052-tendermint-mode.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        - mempool, state
```
Comment (marbar3778) [inline docs/architecture/adr-052-tendermint-mode.md]:
```suggestion
```

File: docs/architecture/adr-052-tendermint-mode.md [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        - rpc (safe connections only)
+        - *~~no privValidator(priv_validator_key.json, priv_validator_state.json)~~*
+    - validator
+        - switch, transport
+        - all reactors
```
Comment (marbar3778) [inline docs/architecture/adr-052-tendermint-mode.md]:
```suggestion
        - reactors
          - mempool
          - consensus
          - evidence
          - blockchain
          - p2p/pex
```

File: docs/architecture/adr-052-tendermint-mode.md [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        - mempool, state
```
Comment (marbar3778) [inline docs/architecture/adr-052-tendermint-mode.md]:
```suggestion
```

File: docs/architecture/adr-052-tendermint-mode.md [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+        - rpc
```
Comment (marbar3778) [inline docs/architecture/adr-052-tendermint-mode.md]:
```suggestion
   - rpc (safe connections only)
```

File: docs/architecture/adr-052-tendermint-mode.md [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+        - with privValidator(priv_validator_key.json, priv_validator_state.json)
+    - seed
+        - switch, transport
+        - only pexReactor
```
Comment (marbar3778) [inline docs/architecture/adr-052-tendermint-mode.md]:
```suggestion
        - reactor
           - p2p/pex
```

File: docs/architecture/adr-052-tendermint-mode.md (hunk 5 above)
Comment (marbar3778) [reply docs/architecture/adr-052-tendermint-mode.md]: Yea this is starting to play into a redesign of the rpc (possibly having different levels, that is out of scope) so for now I'd say that is fine.

**Comment by marbar3778** [comment]
can you add this adr to the table of contents in the readme of the adr dir

**Pull request merged**